File: evoHike.Frontend/src/utils/geoUtils.ts
export const calculateGeoJsonLength = (coordinates: number[][]): number => {
  let totalDistance = 0;
  for (let i = 0; i < coordinates.length - 1; i++) {
    const [lon1, lat1] = coordinates[i];
    const [lon2, lat2] = coordinates[i + 1];
    totalDistance += getDistanceFromLatLonInMeters(lat1, lon1, lat2, lon2);
  }
  return totalDistance;
};

const getDistanceFromLatLonInMeters = (
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
) => {
  const R = 6371e3; // Föld sugara méterben
  const dLat = deg2rad(lat2 - lat1);
  const dLon = deg2rad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(deg2rad(lat1)) *
      Math.cos(deg2rad(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
};

const deg2rad = (deg: number) => deg * (Math.PI / 180);

File: evoHike.Backend/Utils/GeoUtils.cs
using NetTopologySuite.Geometries;

namespace evoHike.Backend.Utils;

public static class GeoUtils
{
    private const double EarthRadiusKm = 6371.0;

    public static double CalculateLengthKm(Geometry? geometry)
    {
        if (geometry == null) 
            return 0;

        double totalDistance = 0;

        switch (geometry)
        {
            case LineString ls:
                totalDistance += CalculateLineLength(ls.Coordinates);
                break;
            case MultiLineString mls:
                foreach (var geom in mls.Geometries)
                {
                    if (geom is LineString subLine)
                        totalDistance += CalculateLineLength(subLine.Coordinates);
                }
                break;
            case GeometryCollection gc:
                totalDistance += gc.Geometries.Sum(CalculateLengthKm);
                break;
        }

        return Math.Round(totalDistance, 2);
    }

    private static double CalculateLineLength(Coordinate[] coords)
    {
        double length = 0;
        for (int i = 0; i < coords.Length - 1; i++)
        {
            length += EquirectangularDistance(coords[i], coords[i + 1]);
        }
        return length;
    }

    private static double EquirectangularDistance(Coordinate p1, Coordinate p2)
    {
        var lat1 = ToRadians(p1.Y);
        var lat2 = ToRadians(p2.Y);
        var lon1 = ToRadians(p1.X);
        var lon2 = ToRadians(p2.X);

        var x = (lon2 - lon1) * Math.Cos((lat1 + lat2) / 2);
        var y = lat2 - lat1;

        return Math.Sqrt(x * x + y * y) * EarthRadiusKm;
    }

    private static double ToRadians(double degrees) => degrees * Math.PI / 180.0;
}
